src/spec-node/featuresCLI/packageCommandImpl.ts

```diff
@@ -0,0 +1,84 @@
+import path from 'path';
+import tar from 'tar';
+import { Feature } from '../../spec-configuration/containerFeaturesConfiguration';
+import { LogLevel } from '../../spec-utils/log';
+import { isLocalFile, readLocalDir, readLocalFile, writeLocalFile } from '../../spec-utils/pfs';
+import { FeaturesPackageCommandInput } from './package';
+
+
+export interface SourceInformation {
+    source: string;
+    owner?: string;
+    repo?: string;
+    tag?: string;
+    ref?: string;
+    sha?: string;
+}
+
+export interface DevContainerCollectionMetadata {
+    sourceInformation: SourceInformation;
+    features: Feature[];
+}
+
+
+export async function doFeaturesPackageCommand(args: FeaturesPackageCommandInput): Promise<number> {
+    const { output } = args;
+
+    // For each feature, package each feature and write to 'outputDir/{f}.tgz'
+    // Returns an array of feature metadata from each processed feature
+    const metadataOutput = await getFeaturesAndPackage(args);
+
+    if (!metadataOutput) {
+        output.write('Failed to package features', LogLevel.Error);
+        return 1;
+    }
+
+    const collection: DevContainerCollectionMetadata = {
+        sourceInformation: {
+            source: 'devcontainer-cli',
+        },
+        features: metadataOutput,
+    };
+
+    // Write the metadata to a file
+    const metadataOutputPath = path.join(args.outputDir, 'devcontainer-collection.json');
+    await writeLocalFile(metadataOutputPath, JSON.stringify(collection, null, 4));
+
+    return 0;
+}
+
+async function tarDirectory(featureFolder: string, archiveName: string, outputDir: string) {
+    return new Promise<void>((resolve) => resolve(tar.create({ file: path.join(outputDir, archiveName), cwd: featureFolder }, ['.'])));
+}
+
+export async function getFeaturesAndPackage(args: FeaturesPackageCommandInput): Promise<Feature[] | undefined> {
+    const { output, featuresFolder, outputDir } = args;
+    const featureDirs = await readLocalDir(featuresFolder);
+    let metadatas: Feature[] = [];
+
+    for await (const f of featureDirs) {
+        output.write(`Processing feature: ${f}...`, LogLevel.Info);
+        if (!f.startsWith('.')) {
+            const featureFolder = path.join(featuresFolder, f);
+            const archiveName = `devcontainer-feature-${f}.tgz`;
+
+            await tarDirectory(featureFolder, archiveName, outputDir);
```

Comment:
Can you move this down to below the validations? We should check for the existence of a `devcontainer-feature.json` and an `install.sh` before tarring up the directory. If one of those does not exist, print a warning (probably not an error unless you are trying the single feature case mentioned above?) and skip that folder.